File: LRUCache.py
class LRUCache:

    def __init__(self, capacity):
        self.cache = {}
        self.capacity = capacity
        self.head = None
        self.tail = None
    
    def put(self, key, value):
        if key in self.cache:
            #update ll
            # if not only one node 
            if(self.head != self.tail):
                node = self.cache[key]
                prev = node.prev 
                next = node.next
                node.prev = None
                node.next = None
                prev.next = next
                if next != None: 
                    next.prev = prev
                node.next = self.head
                self.head.prev = node
                self.head = node   
                print("cache", self.cache)
                self.printLL()         
        else:   
            if len(self.cache) == self.capacity:
                #remove tail from ll
                prev = self.tail.prev
                prev.next = None
                self.tail = prev
                #remove the item from cache 
                del self.cache[key]
                #add this new node to head 
                node = Node(key, value)
                node.next = self.head
                self.head.prev = node
                self.head = node
                #add this item to cache
                self.cache[key] = node
            else:
                #add to cache 
                node = Node(key, value)
                self.cache[key] = node
                #add to head of ll
                if self.head is not None:
                    node.next = self.head
                    self.head.prev = node
                    self.head = node
                else :
                    pass

    def get(self, key):
        if key in self.cache:
            node = self.cache[key]
            prev = node.prev 
            next = node.next
            node.prev = None
            node.next = None
            prev.next = next
            if next != None: 
                next.prev = prev
            node.next = self.head
            self.head.prev = node
            self.head = node
            return node.val
        else: 
            return -1

    def printLL(self):
        node = self.head
        while(node is not None):
            print(node.key)
            node = node.next

class Node:
    def __init__(self, key, val):
        self.key = key
        self.val = val
        self.prev = None
        self.next = None
        pass

cache = LRUCache(2)
print(cache.put(1,1))
print(cache.put(2,2))
print(cache.get(1))
print(cache.put(3,3))
print(cache.get(2))
print(cache.put(4,4))
print(cache.get(1))
print(cache.get(3))
print(cache.get(4))

File: LRUCache.java
import java.util.HashMap;
import java.util.LinkedList;
import java.util.Map;

/**
 * cache capacity
 * cache
 *
 * cap - 4
 * A
 * A B
 * A B C
 * A B C D
 */

class LRUCache {

    Map<Integer, Node> cache;
    int capacity;
    Node head = null;
    Node tail = null;

    public LRUCache(int capacity) {
        this.cache = new HashMap<>(capacity);
        this.capacity = capacity;
    }

    public int get(int key) {
        if(cache.containsKey(key)){
            Node node = cache.get(key);
            Node prev = node.prev;
            prev.next = node.next;
            node.next.prev = prev;
            node.next = head;
            head.prev = node;
            head = node;
            return cache.get(key).val;
        } else return -1;
    }

    public void put(int key, int value) {
        if(cache.containsKey(key)){
            // update the ll
            Node node = cache.get(key);
            Node prev = node.prev;
            prev.next = node.next;
            node.next.prev = prev;
            node.next = head;
            head.prev = node;
            head = node;
        } else {
            if(cache.size() == capacity){
                // decide which to evict then evict & insert
                int keyToRemove = tail.key;
                tail = tail.prev;
                tail.next = null;
                cache.remove(keyToRemove);
                Node node = new Node(key, value);
                cache.put(key, node);
                node.next = head;
                head.prev = node;
                head = node;
            } else {
                // add to cache
                Node node = new Node(key, value);
                if(head == null){
                    head = node;
                    tail = node;
                } else {
                    head.prev = node;
                    node.next = head;
                    head = node;
                }
                cache.put(key, node);
            }
        }
    }
}

class Node{
    int key;
    int val;
    Node prev;
    Node next;
    Node(int key, int val){
        this.key = key;
        this.val = val;
        this.prev = null;
        this.next = null;
    }
}

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache obj = new LRUCache(capacity);
 * int param_1 = obj.get(key);
 * obj.put(key,value);
 */
